### ISSUE TITLE
derivable_impls does not trigger on manual Default impl on enum if the impl uses `Self` instead of type name

### ISSUE THREAD

**Issue by Arnavion** [rust-lang/rust-clippy#11897]
### Summary

Described below.

### Lint Name

derivable_impls

### Reproducer

I tried this code:

```rust
#![deny(clippy::derivable_impls)]

pub enum Foo {
    A,
    B,
}

impl Default for Foo {
    fn default() -> Self {
        Self::A
    }
}
```

I expected to see this happen:

`clippy::derivable_impls` triggers and tells me to instead write:

```rust
#[derive(Default)]
pub enum Foo {
    #[default] A,
    B,
}
```

Instead, this happened:

Lint does not trigger.

If I change the impl to return `Foo::A` instead of `Self::A`, then it does trigger:

```rust
impl Default for Foo {
    fn default() -> Self {
        Foo::A
    }
}
```

### Version

```text
rustc 1.76.0-nightly (5facb422f 2023-11-28)
binary: rustc
commit-hash: 5facb422f8a5a61df515572fe79b02433639d565
commit-date: 2023-11-28
host: x86_64-unknown-linux-gnu
release: 1.76.0-nightly
LLVM version: 17.0.5

clippy 0.1.76 (5facb42 2023-11-28)
```

**Comment by Arnavion** [comment]
This was fixed at some point.

```
    Checking playground v0.0.1 (/playground)
error: this `impl` can be derived
  --> src/lib.rs:8:1
   |
 8 | / impl Default for Foo {
 9 | |     fn default() -> Self {
10 | |         Self::A
11 | |     }
12 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
 1 | #![deny(clippy::derivable_impls)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^
help: replace the manual implementation with a derive attribute and mark the default variant
   |
 3 + #[derive(Default)]
 4 | pub enum Foo {
 5 ~     #[default]
 6 ~     A,
   |

error: could not compile `playground` (lib) due to 1 previous error
```
